File: src/main/java/hu/frontrider/arcana/content/items/formula/text/ModifierText.java
package hu.frontrider.arcana.content.items.formula.text;

import hu.frontrider.arcana.core.creatureenchant.EnchantingBaseCircle;
import hu.frontrider.arcana.util.AspectUtil;
import hu.frontrider.arcana.util.StringUtil;
import net.minecraft.client.resources.I18n;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.world.World;
import net.minecraftforge.fml.common.registry.GameRegistry;
import net.minecraftforge.registries.IForgeRegistry;
import thaumcraft.api.aspects.AspectList;

import javax.annotation.Nullable;
import java.util.List;


/**
 * @author Kis András Gábor
 * 2018.08.28.
 */
public class ModifierText implements FormulaText {

    private IForgeRegistry<EnchantingBaseCircle> creatureEnchantIForgeRegistry;

    public ModifierText() {
        creatureEnchantIForgeRegistry = GameRegistry.findRegistry(EnchantingBaseCircle.class);
    }

    @Override
    public void addText(ItemStack stack, NBTTagCompound tagCompound, @Nullable World worldIn, List<String> tooltip) {
        AspectList aspectList = new AspectList();
        aspectList.readFromNBT(tagCompound);
        StringUtil.INSTANCE.swappedDescription(tooltip)
                .permanent(description -> creatureEnchantIForgeRegistry
                        .getValuesCollection()
                        .forEach(modifier -> {
                            if (AspectUtil.INSTANCE.aspectListEquals(aspectList, modifier.getFormula())) {
                                description.add(I18n.format("enchant.creature_enchant.helper.modifier") + ": " + I18n.format(modifier.getUnlocalizedName()));
                            }
                        }))
                .render();
    }
}

File: src/main/java/hu/frontrider/arcana/sided/client/gui/formula/text/ModifierText.kt
package hu.frontrider.arcana.sided.client.gui.formula.text

import hu.frontrider.arcana.core.creatureenchant.EnchantingBaseCircle
import hu.frontrider.arcana.util.AspectUtil
import hu.frontrider.arcana.util.strings.ChatFormat
import hu.frontrider.arcana.util.strings.StringUtil
import hu.frontrider.arcana.util.strings.formatTranslate
import net.minecraft.client.resources.I18n
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.world.World
import net.minecraftforge.fml.common.registry.GameRegistry
import net.minecraftforge.registries.IForgeRegistry
import thaumcraft.api.aspects.AspectList


/**
 * @author Kis András Gábor
 * 2018.08.28.
 */
class ModifierText : FormulaText {

    private val creatureEnchantIForgeRegistry: IForgeRegistry<EnchantingBaseCircle>

    init {
        creatureEnchantIForgeRegistry = GameRegistry.findRegistry(EnchantingBaseCircle::class.java)
    }

    override fun addText(stack: ItemStack, tagCompound: NBTTagCompound, worldIn: World?, tooltip: MutableList<String>) {
        val aspectList = AspectList()
        aspectList.readFromNBT(tagCompound)
                    creatureEnchantIForgeRegistry
                            .valuesCollection
                            .forEach { modifier ->
                                if (AspectUtil.aspectListEquals(aspectList, modifier.formula)) {
                                    tooltip+=("${formatTranslate("enchant.creature_enchant.helper.modifier", ChatFormat.YELLOW)}: ${formatTranslate(modifier.unlocalizedName, ChatFormat.YELLOW)}")
                                }
                            }

    }
}
